pkg/apis/core/validation/validation.go

```diff
@@ -1781,12 +1779,9 @@ func ValidatePersistentVolumeUpdate(newPv, oldPv *core.PersistentVolume) field.E
 	if !apiequality.Semantic.DeepEqual(newPv.Spec.PersistentVolumeSource, oldPv.Spec.PersistentVolumeSource) {
 		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "persistentvolumesource"), "is immutable after creation"))
 	}
-
 	newPv.Status = oldPv.Status
 
-	if utilfeature.DefaultFeatureGate.Enabled(features.BlockVolume) {
-		allErrs = append(allErrs, ValidateImmutableField(newPv.Spec.VolumeMode, oldPv.Spec.VolumeMode, field.NewPath("volumeMode"))...)
-	}
+	allErrs = append(allErrs, ValidateImmutableField(newPv.Spec.VolumeMode, oldPv.Spec.VolumeMode, field.NewPath("volumeMode"))...)
```

Comment:
This introduces a failure mode when reading an existing object with a nil volumeMode field from a server with the gate disabled, then writing to a server with the gate enabled. Here are all the scenarios:

* existing object has nil volumeMode field, read from server with gate enabled, written to server with gate enabled
  * field defaulted to "filesystem" on read (because gate is enabled on read server)
  * client round-trips, submits object with "filesystem"
  * field is left as-is on write (because gate is enabled on write server)
  * oldPv field is defaulted to "filesystem" in write server (because gate is enabled on write server)
  * immutable check PASSES, oldPv and newPV both have "filesystem"
* existing object has nil volumeMode field, read from server with gate enabled, written to server with gate disabled
  * field defaulted to "filesystem" on read (because gate is enabled on read server)
  * client round-trips, submits object with "filesystem"
  * field is set to nil on write (because gate is disabled on write server and existing object has a nil field)
  * oldPv field is nil in write server (because gate is disabled on write server)
  * immutable check PASSES, oldPv and newPv both have nil
* existing object has nil volumeMode field, read from server with gate disabled, written to server with gate enabled
  * field is returned as nil on read (because gate is disabled on read server)
  * client round-trips, submits object with nil
  * field is left as-is on write (because feature gate is enabled on write server)
  * oldPv field is defaulted to "filesystem" in write server (because gate is enabled on write server)
  * **immutable check FAILS, oldPv has "filesystem", newPV has nil**
* existing object has nil volumeMode field, read from server with gate disabled, written to server with gate disabled
  * field is returned as nil on read (because gate is disabled on read server)
  * client round-trips, submits object with nil
  * field is set to nil on write (because feature gate is disabled on write server and existing object has a nil field)
  * oldPv field is nil in write server (because gate is disabled on write server)
  * immutable check PASSES, oldPv and newPv both have nil
* existing object has non-nil volumeMode field, read and then written to server with gate enabled
  * existing field is returned on read
  * client round-trips, submits object with same field value
  * field is left as-is on write (because feature gate is enabled on write server)
  * oldPv field is non-nil in write server
  * immutable check PASSES, oldPv and newPv both have identical non-nil values
* existing object with non-nil volumeMode field, read and then written to server with gate disabled
  * existing field is returned on read
  * client round-trips, submits object with same field value
  * field is left as-is on write (because existing object has a non-nil value)
  * oldPv field is non-nil in write server
  * immutable check PASSES, oldPv and newPv both have identical non-nil values

Reply:
given the volume mode feature is beta, the default is locked to filesystem now and cannot be changed. I'd remove the conditional feature gate check in the defaulting code and always default to filesystem. That would fix the immutable check to work in all cases.